File: VB/GridRecentItemsHelper.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.XtraGrid.Views.Grid
Imports System.ComponentModel
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports System.Windows.Forms
Imports System.Drawing
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls

Namespace WindowsApplication1
	Public Class GridRecentItemsHelper

		Private _EditPinned As RepositoryItemButtonEdit
			  Private _EditUnPinned As RepositoryItemButtonEdit
		Public Sub New(ByVal view As GridView)
			_View = view
			view.BeginUpdate()
			InitView()
			InitColumns()
			view.EndUpdate()
		End Sub

		Private Sub InitView()
			_View.GridControl.ForceInitialize()
			_View.GridControl.DataSource = RecentItems
			_View.Columns("IsPinned").GroupIndex = 0
			_View.Columns("IsPinned").SortOrder = DevExpress.Data.ColumnSortOrder.Descending
			_View.OptionsView.ShowGroupPanel = False
			_View.OptionsView.ShowHorzLines = False
			_View.OptionsView.GroupDrawMode = GroupDrawMode.Office2003
			_View.GroupRowHeight = 20
			_View.OptionsView.ShowVertLines = False
			_View.OptionsBehavior.AutoExpandAllGroups = True
			_View.OptionsBehavior.ImmediateUpdateRowPosition = True
			_View.OptionsView.ShowColumnHeaders = False
			_View.OptionsSelection.EnableAppearanceFocusedCell = False
			_View.FocusRectStyle = DrawFocusRectStyle.None
			AddHandler _View.CustomDrawGroupRow, AddressOf view_CustomDrawGroupRow
			AddHandler _View.CustomDrawCell, AddressOf view_CustomDrawCell
			AddHandler _View.GroupRowCollapsing, AddressOf view_GroupRowCollapsing
			AddHandler _View.MouseMove, AddressOf view_MouseMove
			AddHandler _View.CustomRowCellEdit, AddressOf _View_CustomRowCellEdit
			_View.OptionsView.ShowIndicator = False
		End Sub

		Private Sub _View_CustomRowCellEdit(ByVal sender As Object, ByVal e As CustomRowCellEditEventArgs)
			If e.Column.FieldName = "PinnedImage" AndAlso e.RowHandle = _View.FocusedRowHandle Then
				e.RepositoryItem = If(True.Equals(e.CellValue), _EditPinned, _EditUnPinned)
			End If
		End Sub
		Private Sub InitColumns()
			Dim col As DevExpress.XtraGrid.Columns.GridColumn = _View.Columns("Image")
			col.OptionsColumn.AllowEdit = False
			col.VisibleIndex = 0
			col.MinWidth = 16
			col.Width = 10
			col = _View.Columns("Caption")
			col.OptionsColumn.AllowEdit = False
			col = _View.Columns("PinnedImage")
			col.Width = 16
			_EditPinned = New RepositoryItemButtonEdit()
			_EditPinned.TextEditStyle = TextEditStyles.HideTextEditor
			_EditPinned.Buttons(0).Kind = ButtonPredefines.Glyph
			AddHandler _EditPinned.ButtonClick, AddressOf _EditPinned_ButtonClick
			_EditUnPinned = TryCast(_EditPinned.Clone(), RepositoryItemButtonEdit)
			_EditPinned.Buttons(0).Image = My.Resources.pin16
			_EditUnPinned.Buttons(0).Image = My.Resources.unpin16
		End Sub

		Private Sub _EditPinned_ButtonClick(ByVal sender As Object, ByVal e As ButtonPressedEventArgs)
			_View.CloseEditor()
			Dim item As GridRecentItem = TryCast(_View.GetFocusedRow(), GridRecentItem)
			item.IsPinned = Not item.IsPinned
			_View.RefreshData()
			_View.ExpandAllGroups()
		End Sub

		Private Sub view_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim hi As GridHitInfo = _View.CalcHitInfo(e.Location)
			If hi.InRow Then
				_View.FocusedRowHandle = hi.RowHandle
			End If
		End Sub



		Private Sub view_GroupRowCollapsing(ByVal sender As Object, ByVal e As RowAllowEventArgs)
			e.Allow = False
		End Sub

		Private Sub DrawPinnedImage(ByVal e As RowCellCustomDrawEventArgs)
			If e.RowHandle <> _View.FocusedRowHandle Then
				e.Handled = True
			End If

		End Sub

		Private Sub view_CustomDrawCell(ByVal sender As Object, ByVal e As RowCellCustomDrawEventArgs)
			DrawHotTrack(e)
			Select Case e.Column.FieldName
				Case "Image"
					DrawImage(e)
				Case "Caption"
					DrawCaption(e)
				Case "PinnedImage"
					DrawPinnedImage(e)
				Case Else
			End Select
		End Sub

		Private Sub DrawHotTrack(ByVal e As RowCellCustomDrawEventArgs)

		End Sub

		Private Sub DrawImage(ByVal e As RowCellCustomDrawEventArgs)
			Dim image As Image = TryCast(e.CellValue, Image)
			If image Is Nothing Then
				Return
			End If
			Dim bounds As Rectangle = e.Bounds
			bounds.Inflate(-2, -2)
			e.Cache.Paint.DrawImage(e.Cache.Graphics, image, bounds)
			e.Handled = True
		End Sub
		Private Sub DrawCaption(ByVal e As RowCellCustomDrawEventArgs)
			Dim bounds As Rectangle = e.Bounds
			bounds.Inflate(-2, -2)
			e.Appearance.DrawString(e.Cache, e.DisplayText, bounds)
			e.Handled = True
		End Sub
		Private Sub view_CustomDrawGroupRow(ByVal sender As Object, ByVal e As RowObjectCustomDrawEventArgs)
			Dim info As GridGroupRowInfo = CType(CType(e, CustomDrawObjectEventArgs).Info, GridGroupRowInfo)
			Dim bounds As Rectangle = info.DataBounds
			bounds.Offset(2, 0)
			Dim text As String = If(DirectCast(info.EditValue, Boolean), "Pinned", "Recent")
			e.Appearance.DrawString(e.Cache, text, bounds)
			Dim size As SizeF = e.Appearance.CalcTextSize(e.Cache, text, bounds.Width)
			bounds.Offset(size.ToSize().Width, 0)
			bounds.Width -= size.ToSize().Width
			bounds.Inflate(-2, 0)
			Dim point1 As New Point(bounds.X, bounds.Y + bounds.Height \ 2)
			Dim point2 As New Point(bounds.Right, bounds.Y + bounds.Height \ 2)
			e.Cache.DrawLine(e.Cache.GetPen(e.Appearance.GetForeColor()), point1, point2)
			e.Handled = True
		End Sub

		Private _RecentItems As New BindingList(Of GridRecentItem)()
		Private ReadOnly _View As GridView
		Public Property RecentItems() As BindingList(Of GridRecentItem)
			Get
				Return _RecentItems
			End Get
			Set(ByVal value As BindingList(Of GridRecentItem))
				_RecentItems = value
			End Set
		End Property
	End Class
End Namespace

File: CS/GridRecentItemsHelper.cs
using System;
using System.Collections.Generic;
using DevExpress.XtraGrid.Views.Grid;
using System.ComponentModel;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System.Windows.Forms;
using System.Drawing;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;

namespace WindowsApplication1
{
    public class GridRecentItemsHelper
    {

        private RepositoryItemButtonEdit _EditPinned;
              private RepositoryItemButtonEdit _EditUnPinned;
        public GridRecentItemsHelper(GridView view)
        {
            _View = view;
            view.BeginUpdate();
            InitView();
            InitColumns();
            view.EndUpdate();
        }

        private void InitView()
        {
            _View.GridControl.ForceInitialize();
            _View.GridControl.DataSource = RecentItems;
            _View.Columns["IsPinned"].GroupIndex = 0;
            _View.Columns["IsPinned"].SortOrder = DevExpress.Data.ColumnSortOrder.Descending;
            _View.OptionsView.ShowGroupPanel = false;
            _View.OptionsView.ShowHorzLines = false;
            _View.OptionsView.GroupDrawMode = GroupDrawMode.Office2003;
            _View.GroupRowHeight = 20;
            _View.OptionsView.ShowVertLines = false;
            _View.OptionsBehavior.AutoExpandAllGroups = true;
            _View.OptionsBehavior.ImmediateUpdateRowPosition = true;
            _View.OptionsView.ShowColumnHeaders = false;
            _View.OptionsSelection.EnableAppearanceFocusedCell = false;
            _View.FocusRectStyle = DrawFocusRectStyle.None;
            _View.CustomDrawGroupRow += view_CustomDrawGroupRow;
            _View.CustomDrawCell += view_CustomDrawCell;
            _View.GroupRowCollapsing += view_GroupRowCollapsing;
            _View.MouseMove += view_MouseMove;
            _View.CustomRowCellEdit += _View_CustomRowCellEdit;
            _View.OptionsView.ShowIndicator = false;
        }

        void _View_CustomRowCellEdit(object sender, CustomRowCellEditEventArgs e)
        {
            if (e.Column.FieldName == "PinnedImage" && e.RowHandle == _View.FocusedRowHandle)
                e.RepositoryItem = true.Equals(e.CellValue) ? _EditPinned : _EditUnPinned ;
        }
        private void InitColumns()
        {
            DevExpress.XtraGrid.Columns.GridColumn col = _View.Columns["Image"];
            col.OptionsColumn.AllowEdit = false;
            col.VisibleIndex = 0;
            col.MinWidth = 16;
            col.Width = 10;
            col = _View.Columns["Caption"];
            col.OptionsColumn.AllowEdit = false;
            col = _View.Columns["PinnedImage"];
            col.Width = 16;
            _EditPinned = new RepositoryItemButtonEdit();
            _EditPinned.TextEditStyle = TextEditStyles.HideTextEditor;
            _EditPinned.Buttons[0].Kind = ButtonPredefines.Glyph;
            _EditPinned.ButtonClick += _EditPinned_ButtonClick;
            _EditUnPinned = _EditPinned.Clone() as RepositoryItemButtonEdit;
            _EditPinned.Buttons[0].Image = WindowsApplication1.Properties.Resources.pin16;
            _EditUnPinned.Buttons[0].Image = WindowsApplication1.Properties.Resources.unpin16;
        }

        void _EditPinned_ButtonClick(object sender, ButtonPressedEventArgs e)
        {
            _View.CloseEditor();
            GridRecentItem item = _View.GetFocusedRow() as GridRecentItem;
            item.IsPinned = !item.IsPinned;
            _View.RefreshData();
            _View.ExpandAllGroups();
        }

        void view_MouseMove(object sender, MouseEventArgs e)
        {
            GridHitInfo hi = _View.CalcHitInfo(e.Location);
            if (hi.InRow)
                _View.FocusedRowHandle = hi.RowHandle;
        }

        

        void view_GroupRowCollapsing(object sender, RowAllowEventArgs e)
        {
            e.Allow = false;
        }

        private void DrawPinnedImage(RowCellCustomDrawEventArgs e)
        {
            if (e.RowHandle != _View.FocusedRowHandle)
                e.Handled = true;

        }

        void view_CustomDrawCell(object sender, RowCellCustomDrawEventArgs e)
        {
            DrawHotTrack(e);
            switch (e.Column.FieldName)
            {
                case "Image":
                    DrawImage(e);
                    break;
                case "Caption":
                    DrawCaption(e);
                    break;
                case "PinnedImage":
                    DrawPinnedImage(e);
                    break;
                default:
                    break;
            }
        }

        private void DrawHotTrack(RowCellCustomDrawEventArgs e)
        {
            
        }

        private void DrawImage(RowCellCustomDrawEventArgs e)
        {
            Image image = e.CellValue as Image;
            if (image == null)
                return;
            Rectangle bounds = e.Bounds;
            bounds.Inflate(-2, -2);
            e.Graphics.DrawImage(image, bounds);
            e.Handled = true;
        }
        private void DrawCaption(RowCellCustomDrawEventArgs e)
        {
            Rectangle bounds = e.Bounds;
            bounds.Inflate(-2, -2);
            e.Appearance.DrawString(e.Cache, e.DisplayText, bounds);
            e.Handled = true;
        }
        void view_CustomDrawGroupRow(object sender, RowObjectCustomDrawEventArgs e)
        {
            GridGroupRowInfo info = (GridGroupRowInfo)(((CustomDrawObjectEventArgs)e).Info);
            Rectangle bounds = info.DataBounds;
            bounds.Offset(2, 0);
            string text = (bool)info.EditValue ? "Pinned" : "Recent";
            e.Appearance.DrawString(e.Cache, text, bounds);
            SizeF size = e.Appearance.CalcTextSize(e.Cache, text, bounds.Width);
            bounds.Offset(size.ToSize().Width, 0);
            bounds.Width -= size.ToSize().Width;
            bounds.Inflate(-2, 0);
            Point point1 = new Point(bounds.X, bounds.Y + bounds.Height / 2);
            Point point2 = new Point(bounds.Right, bounds.Y + bounds.Height / 2);
            e.Graphics.DrawLine(new Pen(e.Appearance.GetForeBrush(e.Cache)), point1, point2);
            e.Handled = true;
        }

        private BindingList<GridRecentItem> _RecentItems = new BindingList<GridRecentItem>();
        private readonly GridView _View;
        public BindingList<GridRecentItem> RecentItems
        {
            get { return _RecentItems; }
            set { _RecentItems = value; }
        }
    }
}
